calico.go

```diff
@@ -0,0 +1,302 @@
+			fmt.Fprintf(os.Stderr, "Calico CNI appending networking profile=%s\n", profileID)
+			theEndpoint.ProfileIDs = append(theEndpoint.ProfileIDs, profileID)
+			result, err = createResultFromIP(theEndpoint.IPv4Nets[0])
+			if err != nil {
+				return err
+			}
+		} else {
+			// There's no existing endpoint, so we need to do the following:
+			// 1) Call the configured IPAM plugin to get IP address(es)
+			// 2) Configure the calico endpoint
+			// 3) Create the veth, configuring it on both the host and container namespace.
+
+			// 1) run the IPAM plugin and make sure there's an IP address returned.
+			result, err = ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
+			if err != nil {
+				return err
+			}
+
+			if result.IP4 == nil && result.IP6 == nil {
+				return errors.New("IPAM plugin did not return any IP addresses")
+			}
+
+			// 2) Create the endpoint object
+			theEndpoint = &libcalico.Endpoint{}
+			theEndpoint.Labels = map[string]string{}
+			theEndpoint.ProfileIDs = []string{profileID}
+			theEndpoint.OrchestratorID = orchestratorID
+			theEndpoint.WorkloadID = workloadID
+			theEndpoint.Hostname = hostname
+			theEndpoint.Mac = "EE:EE:EE:EE:EE:EE"
+			theEndpoint.State = "active"
+
+			if result.IP4 != nil {
+				result.IP4.IP.Mask = net.CIDRMask(32, 32)
+				theEndpoint.IPv4Nets = []string{result.IP4.IP.String()}
+			} else {
+				theEndpoint.IPv4Nets = []string{}
+			}
+
+			if result.IP6 != nil {
+				result.IP6.IP.Mask = net.CIDRMask(128, 128)
+				theEndpoint.IPv6Nets = []string{result.IP6.IP.String()}
+			} else {
+				theEndpoint.IPv6Nets = []string{}
+			}
+			fmt.Fprintf(os.Stderr, "Calico CNI using IPv4=%s IPv6=%s\n", theEndpoint.IPv4Nets, theEndpoint.IPv6Nets)
+
+			// 3) Set up the veth
+			hostVethName, contVethMac, err := DoNetworking(args, conf, result)
+			if err != nil {
+				return err
+			}
+			theEndpoint.Mac = contVethMac
+			theEndpoint.Name = hostVethName
+		}
+
+		// Write the endpoint object (either the newly created one, or the updated one with a new ProfileIDs).
+		if err := theEndpoint.Write(etcd); err != nil {
+			return err
+		}
+	}
+
+	// Handle profile creation.
+	// If Kubernetes is being used then profiles only need to be created if there is no policy block in the network
+	// config. If there is a policy block then "proper" policy is being used and the policy controller handles
+	// profile creation.
+	if !runningUnderK8s || conf.Policy == nil {
+		//TODO - this is the wrong test. It should be on the policy type
+		// Start by checking if the profile already exists. If it already exists then there is no work to do (the CNI plugin never updates a profile).
+		exists, err := libcalico.ProfileExists(conf.Name, etcd)
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			// The profile doesn't exist so needs to be created. The rules vary depending on whether k8s is being used.
+			// Under k8s (without full policy support) the rule is permissive and allows all traffic.
```

Comment:
I'd be interested in getting your opinion on this - I wonder if the CNI plugin should actually be making this judgement call.  Instead, users could just configure the profile as part of the install step (this makes it more clear what is going on and gives some more flexibility in terms of what the default policy is, it also means the profile processing for Kubernetes vs non-Kubernetes is the same).

I don't think this would affect any existing deployments.